File: R_workflow/tests/testthat/test-paradigms.R
LEX <- c("cat", "cap", "car", "can", "cab", "bat", "bad", "bag", "ban",
         "rat", "ran", "ram", "rag", "tan", "tap", "tar", "mat", "map", "man")

test_that("every paradigm resolves to typed events with its required fields", {
  for (name in names(PARADIGMS)) {
    events <- resolve_events(list(paradigm = name))
    expect_gt(length(events), 0)
    expect_true(all(vapply(events, function(e) !is.null(e$type), logical(1))))
    expect_true(all(PARADIGMS[[name]]$stimulus_fields %in% required_fields(list(paradigm = name))))
  }
})

test_that("content_field and referenced_fields read field references", {
  expect_identical(content_field("{target}"), "target")
  expect_null(content_field("+"))
  expect_identical(referenced_fields(resolve_events(list(paradigm = "priming"))),
                   c("prime", "target"))
})

test_that("explicit events override the paradigm default", {
  design <- list(events = list(list(type = "text", content = "{w}", duration_frames = 10L)))
  expect_identical(vapply(resolve_events(design), function(e) e$type, character(1)), "text")
})

test_that("pseudowords are deterministic", {
  a <- generate_pseudowords(c("cat", "bat", "rat"), LEX)
  b <- generate_pseudowords(c("cat", "bat", "rat"), LEX)
  expect_identical(a$pseudoword, b$pseudoword)
})

test_that("pseudowords are legal non-words of matched length", {
  gen <- generate_pseudowords(c("cat", "bat", "rat", "man"), LEX)
  bg <- names(bigram_counts(LEX))
  for (i in seq_len(nrow(gen))) {
    pw <- gen$pseudoword[i]
    expect_false(pw %in% LEX)
    expect_equal(nchar(pw), nchar(gen$base_word[i]))
    bgs <- substring(pw, 1:(nchar(pw) - 1L), 2:nchar(pw))
    expect_true(all(bgs %in% bg))
  }
  expect_equal(length(unique(gen$pseudoword)), nrow(gen))
})

test_that("build_lexdec_stimuli pairs words and pseudowords by length", {
  schema <- yaml::read_yaml(system.file("extdata", "schema.yaml", package = "lexsync"))
  lex <- load_lexicon(system.file("extdata", "en_example.csv", package = "lexsync"), schema, "english")
  pool <- build_pool(lex, list(length = c(4, 6), frequency = c(3.0, 6.5)))
  stim <- build_lexdec_stimuli(pool, 15L, reference_words = lex$word)
  expect_setequal(unique(stim$condition), c("word", "pseudoword"))
  expect_equal(sum(stim$condition == "word"), sum(stim$condition == "pseudoword"))
  for (s in unique(stim$set)) expect_equal(length(unique(stim$length[stim$set == s])), 1L)
})

test_that("load_items validates columns and clean_field rejects control characters", {
  p <- tempfile(fileext = ".csv")
  writeLines(c("item,prime", "1,nurse"), p)
  expect_error(load_items(p, c("prime", "target")))
  expect_error(clean_field("ab\ncd", "target"))
  expect_error(clean_field("ab\tcd", "target"))
  expect_identical(clean_field("a clean, quoted 'item'", "x"), "a clean, quoted 'item'")
})

test_that("the Latin square shows each item once per list, balanced", {
  items <- data.frame(
    item = c(1, 1, 2, 2, 3, 3, 4, 4),
    condition = rep(c("related", "unrelated"), 4),
    prime = letters[1:8],
    target = c("x", "x", "y", "y", "z", "z", "w", "w"),
    stringsAsFactors = FALSE)
  items$set <- items$item
  out <- counterbalance(items, list(paradigm = "priming", counterbalance = list(lists = 2)),
                        list(seed = 1))
  for (l in unique(out$list)) {
    lst <- out[out$list == l, ]
    expect_equal(length(unique(lst$set)), nrow(lst))
    expect_equal(length(unique(lst$target)), nrow(lst))
    expect_setequal(unique(lst$condition), c("related", "unrelated"))
    expect_equal(sum(lst$condition == "related"), sum(lst$condition == "unrelated"))
  }
})

File: python_workflow/tests/test_paradigms.py
"""Trial-event model, pseudoword generation, item tables and counterbalancing."""
import pandas as pd
import pytest

from lexsync.counterbalancing import counterbalance
from lexsync.generation import (bigram_counts, build_lexdec_stimuli,
                                generate_pseudowords)
from lexsync.io_utils import clean_field
from lexsync.paradigms import (PARADIGMS, content_field, referenced_fields,
                               required_fields, resolve_events)
from lexsync.querying import build_pool, load_items, load_lexicon


# ---- the trial-event model ------------------------------------------------

def test_every_paradigm_resolves_to_typed_events():
    for name in PARADIGMS:
        events = resolve_events({"paradigm": name})
        assert events and all("type" in e for e in events)
        # required fields are exactly the paradigm fields plus any referenced ones
        assert set(PARADIGMS[name]["stimulus_fields"]) <= set(required_fields({"paradigm": name}))


def test_content_field_and_referenced_fields():
    assert content_field("{target}") == "target"
    assert content_field("+") is None
    events = resolve_events({"paradigm": "priming"})
    assert referenced_fields(events) == ["prime", "target"]


def test_explicit_events_override_paradigm():
    design = {"events": [{"type": "text", "content": "{w}", "duration_frames": 10}]}
    assert [e["type"] for e in resolve_events(design)] == ["text"]


# ---- pseudoword generation ------------------------------------------------

LEX = ["cat", "cap", "car", "can", "cab", "bat", "bad", "bag", "ban",
       "rat", "ran", "ram", "rag", "tan", "tap", "tar", "mat", "map", "man"]


def test_pseudowords_are_deterministic():
    a = generate_pseudowords(["cat", "bat", "rat"], LEX)
    b = generate_pseudowords(["cat", "bat", "rat"], LEX)
    assert list(a["pseudoword"]) == list(b["pseudoword"])


def test_pseudowords_are_legal_nonwords_of_matched_length():
    gen = generate_pseudowords(["cat", "bat", "rat", "man"], LEX)
    lex = set(LEX)
    for base, pw in zip(gen["base_word"], gen["pseudoword"], strict=True):
        assert pw not in lex                    # not a real word
        assert len(pw) == len(base)             # length preserved
        bg = bigram_counts(LEX)
        assert all(pw[i:i + 2] in bg for i in range(len(pw) - 1))  # orthographically legal
    assert len(set(gen["pseudoword"])) == len(gen)  # unique


def test_build_lexdec_stimuli_pairs_words_and_pseudowords(schema, en_lexicon_path):
    lex = load_lexicon(en_lexicon_path, schema, "english")
    pool = build_pool(lex, {"length": [4, 6], "frequency": [3.0, 6.5]})
    stim = build_lexdec_stimuli(pool, 15, reference_words=lex["word"].tolist())
    assert set(stim["condition"]) == {"word", "pseudoword"}
    assert (stim["condition"] == "word").sum() == (stim["condition"] == "pseudoword").sum()
    # paired by set, length matched within each set
    for _, g in stim.groupby("set"):
        assert g["length"].nunique() == 1


# ---- item tables and counterbalancing -------------------------------------

def test_load_items_validates_columns(tmp_path):
    p = tmp_path / "bad.csv"
    p.write_text("item,prime\n1,nurse\n", encoding="utf-8")  # missing 'condition'
    with pytest.raises(ValueError):
        load_items(str(p), ["prime", "target"])


def test_clean_field_rejects_control_characters():
    with pytest.raises(ValueError):
        clean_field("ab\ncd", "target")          # newline
    with pytest.raises(ValueError):
        clean_field("ab\tcd", "target")          # tab
    assert clean_field("a clean, quoted 'item'", "x") == "a clean, quoted 'item'"


def test_load_items_rejects_crafted_item(tmp_path):
    p = tmp_path / "items.csv"
    p.write_text('item,condition,target\n1,a,"bad\ttab"\n', encoding="utf-8")
    with pytest.raises(ValueError):
        load_items(str(p), ["target"])


def test_latin_square_shows_each_item_once_per_list_balanced():
    items = pd.DataFrame({
        "item": [1, 1, 2, 2, 3, 3, 4, 4],
        "condition": ["related", "unrelated"] * 4,
        "prime": list("abcdefgh"), "target": ["x", "x", "y", "y", "z", "z", "w", "w"],
    })
    items["set"] = items["item"]
    design = {"paradigm": "priming", "counterbalance": {"lists": 2}}
    out = counterbalance(items, design, {"seed": 1})
    for _, lst in out.groupby("list"):
        assert lst["set"].nunique() == len(lst)              # each item once
        assert lst["target"].nunique() == len(lst)           # no target repeated
        assert set(lst["condition"]) == {"related", "unrelated"}  # both conditions present
        assert (lst["condition"] == "related").sum() == (lst["condition"] == "unrelated").sum()
